src/index.js

```diff
@@ -22,32 +23,24 @@ export default class ScrollBehavior {
     this._stateStorage = stateStorage;
     this._getCurrentLocation = getCurrentLocation;
     this._shouldUpdateScroll = shouldUpdateScroll;
+    this._hasSetScrollRestoration = false;
 
     // This helps avoid some jankiness in fighting against the browser's
     // default scroll behavior on `POP` transitions.
     /* istanbul ignore else: Travis browsers all support this */
-    if (
-      'scrollRestoration' in window.history &&
-      // Unfortunately, Safari on iOS freezes for 2-6s after the user swipes to
-      // navigate through history with scrollRestoration being 'manual', so we
-      // need to detect this browser and exclude it from the following code
-      // until this bug is fixed by Apple.
-      !isMobileSafari()
-    ) {
-      this._oldScrollRestoration = window.history.scrollRestoration;
-      try {
-        window.history.scrollRestoration = 'manual';
-
-        // Scroll restoration persists across page reloads. We want to reset
-        // this to the original value, so that we can let the browser handle
-        // restoring the initial scroll position on server-rendered pages.
-        on(window, 'beforeunload', this._restoreScrollRestoration);
-      } catch (e) {
-        this._oldScrollRestoration = null;
+    this._setScrollRestoration();
+
+    // Scroll restoration persists across page reloads. We want to reset
+    // this to the original value, so that we can let the browser handle
+    // restoring the initial scroll position on server-rendered pages.
+    PageLifecycle.addEventListener('statechange', ({ oldState, newState }) => {
+      if (newState === 'frozen' || newState === 'terminated') {
+        this._restoreScrollRestoration();
       }
-    } else {
-      this._oldScrollRestoration = null;
-    }
+      if (oldState === 'frozen') {
+        this._setScrollRestoration();
+      }
```

Comment:
perhaps we can more simply write this as:

```js
if (
  newState === 'terminated' ||
  newState === 'frozen' ||
  newState === 'discarded'
) {
  this._restoreScrollRestoration();
} else {
  this._setScrollRestoration();
}
```

given the check for whether things have already been set, this should never actually cause unnecessary extra changes

otherwise you'd in principle run `_setScrollRestoration` on e.g. a frozen -> discarded transition, which would be odd

Reply:
done